### PR TITLE
Mingw_w64 fix to get cross-compilation working again 

### PR DESCRIPTION
**Pull request by bnikolic**
###### Motivation for this change
These gets cross compilation for win64 targets using mingw32-w64 working again. Includes PR #33110 .

Key change is to disable hardening when compiling windows.mingw_w64 . Also enable ssp and shared for the final cross compilation GCC so that final version can work with normal hardening flags. 

Tested using:

`nix-build /home/bnikolic/oss/nixpkgs -I /home/bnikolic/oss/ -A bzip2 --arg crossSystem '(import <nixpkgs/lib>).systems.examples.mingwW64' -j 10 --cores 50 --show-trace -K -k      `                   

Needs more testing but hopefully  a good starting point

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [X ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by bnikolic** [comment]
cc @Ericson2314

**Comment by Ericson2314** [comment]
Woo thanks! I'm going to do some investigating of the wrapper symlinks first, but get these fixes right after, all today.

File: pkgs/development/compilers/gcc/6/default.nix [hunk 1]
```diff
@@ -155,13 +155,13 @@ let version = "6.4.0";
           "--enable-threads=win32"
           "--enable-sjlj-exceptions"
           "--enable-hash-synchronization"
-          "--disable-libssp"
+          "--enable-libssp"
           "--disable-nls"
           "--with-dwarf2"
           # I think noone uses shared gcc libs in mingw, so we better do the same.
           # In any case, mingw32 g++ linking is broken by default with shared libs,
           # unless adding "-lsupc++" to any linking command. I don't know why.
```
Comment (DavidEGrayson) [inline pkgs/development/compilers/gcc/6/default.nix]: So this old comment about mingw32 g++ linking being broken with shared libs must be wrong.

Comment (Ericson2314) [reply pkgs/development/compilers/gcc/6/default.nix]: @DavidEGrayson or we didn't build any C++ lol.

**Comment by Ericson2314** [comment]
@bnikolic I updated your PR to avoid the binutils changes. The merge commits are so I can directly compare before and after my PR which broke mingw to begin with, which I do in https://hydra.nixos.org/eval/1421871?compare=ericson2314-old-cross-known-good.

**Comment by Ericson2314** [comment]
I'm going to merge this now, and then start after the hardening change to fix the packages that regressed.

**Pull request merged**